aws/resource_aws_s3_bucket_analytics_configuration.go

```diff
@@ -0,0 +1,437 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+)
+
+func resourceAwsS3BucketAnalyticsConfiguration() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsS3BucketAnalyticsConfigurationPut,
+		Read:   resourceAwsS3BucketAnalyticsConfigurationRead,
+		Update: resourceAwsS3BucketAnalyticsConfigurationPut,
+		Delete: resourceAwsS3BucketAnalyticsConfigurationDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"bucket": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"filter": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"prefix": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							AtLeastOneOf: filterAtLeastOneOfKeys,
+						},
+						"tags": schemaWithAtLeastOneOf(
+							tagsSchema(),
+							filterAtLeastOneOfKeys,
+						),
```

Comment:
Are there expectations to use `schemaWithAtLeastOneOf()` elsewhere? Seems like it is not adding more than a direct `AtLeastOneOf: <value>` to the schema. We generally prefer to explicitly declare the schema, to remove indirection for code readers and allow linters to check usage of the composite literals. 

```go
						"prefix": {
							Type:         schema.TypeString,
							Optional:     true,
							AtLeastOneOf: []string{"filter.0.prefix", "filter.0.tags"},
						},
						"tags": {
							Type:         schema.TypeMap,
							Optional:     true,
							Elem:         &schema.Schema{Type: schema.TypeString},
							AtLeastOneOf: []string{"filter.0.prefix", "filter.0.tags"},
						},
```

Reply:
I was thinking in this case of leveraging the existing pre-defined tags schema. Even though it's not complicated, it's consistent across all resources.
Similar to how we have `tagsSchema()`, `tagsSchemaComputed()`, and `tagsSchemaForceNew()`, but making this a wrapper that could extend any existing schema with an `AtLeastOneOf`.